src/main.rs

```diff
@@ -0,0 +1,291 @@
+            }
+        }
+        let mut below = 0i32;
+        if self.below_pts.capacity() > 1 {
+            let i = self.below_pts.iter();
+            for ii in i {
+                if below > *ii {
+                    below = *ii;
+                }
+            }
+        }
+        (above, below)
+    }
+    fn new() -> Hr {
+        let yc = 32768i32;
+        Hr {
+            dc_ema: yc as f64,
+            lp_ema: yc as f64,
+            threshold_ema: yc as f64,
+            n : 0usize,
+            state : 0u8,
+            timer : 0usize,
+            peak_flag : 0u8,
+            wild_flag : 0u8,
+            above_pts : ConstGenericRingBuffer::<i32, ABOVE_SIZE>::new(),
+            below_pts : ConstGenericRingBuffer::<i32, BELOW_SIZE>::new(),
+            last_peak_n : 0usize,
+            hr : 0f64,
+        }
+    }
+    fn help(&self) -> (u32, u32) {
+        (self.dc_ema as u32, self.threshold_ema as u32)
+    }
+}
+
+static SAMPLE_SIGNAL: Signal<CriticalSectionRawMutex, u32> = Signal::new();
+
+type UART = embassy_stm32::usart::UartTx<'static, embassy_stm32::peripherals::USART3, embassy_stm32::peripherals::DMA1_CH1>;
+static UART_INST: StaticCell<UART> = StaticCell::new();
+
+type LED1 = embassy_stm32::gpio::Output<'static, embassy_stm32::peripherals::PB0>;
+static LED1_INST: StaticCell<LED1> = StaticCell::new();
+type LED3 = embassy_stm32::gpio::Output<'static, embassy_stm32::peripherals::PB14>;
+static LED3_INST: StaticCell<LED3> = StaticCell::new();
+
+type BUTTON1 = embassy_stm32::gpio::Input<'static, embassy_stm32::peripherals::PC13>;
+static BUTTON1_INST: StaticCell<BUTTON1> = StaticCell::new();
+
+#[embassy_executor::task]
+async fn process(uart_ref: &'static mut UART,
+                 led1_ref: &'static mut LED1,
+                 led3_ref: &'static mut LED3,
+                 button1_ref: &'static mut BUTTON1) {
+    let mut msg : String<64> = String::new();
+    msg.clear();
+    core::fmt::write(&mut msg, format_args!("Boot\n")).unwrap();
+    _ = (uart_ref).write(msg.as_bytes()).await;
+    let mut hr = Hr::new();
+    let mut n0 = 0usize;
+    let mut count0 = 0u32;
+    loop {
+        let sample = SAMPLE_SIGNAL.wait().await;
+        let lp = button1_ref.get_level() == Level::Low;
+        led3_ref.set_level(if !lp { High } else { Low });
+        let (n, cooked_sample, peak, state) = hr.tick(lp, sample);
+        led1_ref.set_level(if state != 0 { High } else { Low });
+        if DUMP_MODE {
+            msg.clear();
+            // core::fmt::write(&mut msg, format_args!("{} {}\n",  sample, state+peak)).unwrap();
+            core::fmt::write(&mut msg, format_args!("{} {}\n",  cooked_sample, if lp {1} else {0})).unwrap();
+            // core::fmt::write(&mut msg, format_args!("{}\n",  sample)).unwrap();
+            _ = (uart_ref).write(msg.as_bytes()).await;
+            continue;
+        }
+        if peak != 0 {
+            let rate = hr.hr();
+            let (a, b) = hr.above_below();
```

Comment:
comment on the purpose of `above_below`: to give a measure of the signal strength